### PR TITLE
[ENG-735] Update Next.js and Netlify runtime

### PR DESCRIPTION
**Pull request by jimCresswell**
- Update next.js to 14.
- Specify the Netlify nextjs runtime v5.
- Fix some associated type issues.

Once code review and QA are happy can someone please merge this with a squash commit, then let the devs know that it is in? If I'm not here and tweaks are needed, can someone please make them?

Note: the Netlify Next.js runtime v5 is explicitly set as a dependency here, once merged into `main` we need to opt in the v5 at the site level in Netlify, then create another PR to remove the explicit dependency so that the Netlify platform can control the runtime version.

### CODE REVIEW

**Review by jack-skerrett-oak — APPROVED** [review]
Looks good on /teachers

**Review by BTitterington — APPROVED** [review]

File: netlify/edge-functions/subdomain-redirect.ts [hunk 1]
```diff
@@ -59,13 +66,20 @@ async function redirectNetlifySubdomains(
...
+    logRequestComplete();
     return res;
   }
 }
 
 export default redirectNetlifySubdomains;
+
+export const config: Config = {
+  path: "/*",
+  // onError: "bypass",
```
Comment (mantagen) [inline netlify/edge-functions/subdomain-redirect.ts:84]: Can remove this I guess?

File: src/components/GenericPagesComponents/Select/Select.tsx [hunk 2]
```diff
@@ -97,9 +97,13 @@ const SelectSpan = styled(OakSpan)`
   ${ellipsis}
 `;
 
-export function Select<T extends object>(
-  props: AriaSelectProps<T> & SelectProps,
-) {
+export function Select<
+  T extends {
+    // Set type of value as string, is this right?
```
Comment (mantagen) [inline src/components/GenericPagesComponents/Select/Select.tsx:102]: I think if it works and TS doesn't complain that it's likely right. Can remove this comment before merging.

File: src/node-lib/curriculum-api-2023/queries/pupilLesson/pupilLesson.schema.ts [hunk 3]
```diff
@@ -42,7 +42,8 @@ export type ImageOrTextItem = ImageItem | TextItem;
 
 export const lessonContentSchema = lessonContentSchemaFull.omit({
   state: true,
-  cohort: true,
+  // cohort has type never, investigate.
+  // cohort: true,
```
Comment (mantagen) [inline src/node-lib/curriculum-api-2023/queries/pupilLesson/pupilLesson.schema.ts:46]: @benprotheroe do you know if this is required? Can we remove this line?

Comment (benprotheroe) [reply src/node-lib/curriculum-api-2023/queries/pupilLesson/pupilLesson.schema.ts:46]: yeah i dont think you need this line, lessonContentSchemaFull doesnt have a cohort field

File: src/node-lib/cms/sanity-client/parseResults.ts [hunk 4]
```diff
@@ -141,5 +142,6 @@ export const parseResults = <S extends ZodSchema, D>(
     }
   }
 
-  return schema.parse(data);
+  // Had to add a type assertion here as the inferred type was unknown.
+  return schema.parse(data) as ReturnType<S["parse"]>;
```
Comment (mantagen) [inline src/node-lib/cms/sanity-client/parseResults.ts:146]: Have fixed by passing a `O` generic as output

**Review by mantagen — APPROVED** [review]
Code LGTM

**Comment by BTitterington** [comment]
@mantagen - Those changes to parseResults don't seem to play well with check-types. I've spent a little while trying to work your change in and fix things, but the typescript is a little too deep for me in the time available. I'd suggest reverting and creating a ticket to resolve later if we feel strongly about improving this.

File: src/node-lib/cms/sanity-client/parseResults.ts (hunk 4 above)
Comment (mantagen) [reply src/node-lib/cms/sanity-client/parseResults.ts]: Whoops -- hasn't fixed 🤷

**Comment by jimCresswell** [comment]
@BTitterington @mantagen please liaise with Netlify on this, Ian can add you into the appropriate channel

**Comment by BTitterington** [comment]
> @BTitterington @mantagen please liaise with Netlify on this, Ian can add you into the appropriate channel

@jimCresswell I've let them know what happened with the fix they proposed in the channel (it didn't work), we'll see what they come back with.

@mantagen I can add you to the channel also.

**Review by orangemug — APPROVED** [review]

**Comment by jimCresswell** [comment]
This briefly broke production, backing out

**Pull request merged**